iree/compiler/Codegen/Utils/Utils.cpp

```diff
@@ -187,44 +190,32 @@ static SmallVector<Value> getValuesForDimsOrSymbols(
   return vals;
 }
 
-/// Returns the dimension for any operation that has the `dimension`
-/// attribute. Currently tested for `flow.dispatch.workgroup.(size|count|id)`,
-/// `hal.interface.workgroup.(size|count|id)`.
-template <typename T>
-static Optional<unsigned> getDimension(Operation *op) {
-  if (auto tOp = dyn_cast<T>(op)) {
-    return tOp.dimension().getZExtValue();
-  }
-  return llvm::None;
-}
-template <typename T1, typename T2, typename... T3>
-static Optional<unsigned> getDimension(Operation *op) {
-  if (!op) return llvm::None;
-  if (auto dimension = getDimension<T1>(op)) {
-    return dimension;
-  }
-  return getDimension<T2, T3...>(op);
-}
-
-/// Checks that all `vals` are defined by either
-/// `flow.dispatch.workgroup.(size|count|id)` or
-/// `hal.interface.workgroup.(size|count|id)` using the same `dimension`. If any
-/// element of `vals` is not defined by one of these ops, or the dimensions dont
-/// match, returns llvm::None. On success returns the dimension.  If
-/// `refDimension` is passed checks if the dimension matches the given value.
-template <typename... T>
+/// Checks that all `vals` are defined by some processor id/count/size ops using
+/// the same `dimension`. If any element of `vals` is not defined by one of
+/// these ops, or the dimensions dont match, returns llvm::None; oterhwise,
+/// returns the dimension.  If `refDimension` is passed checks if the dimension
+/// matches the given value.
 static Optional<unsigned> checkDimensions(
-    ArrayRef<Value> vals, Optional<unsigned> refDimension = llvm::None) {
+    ArrayRef<Value> vals, bool allowIdOp, bool allowCountOp,
```

Comment:
Not a blocker, I see you explicitly changed this, but I prefer the approach that was there previously. This is too verbose for my liking :)

Reply:
Okay; reverted back to templated impl.